rfcs/20220925-bounded-dynamism.md

```diff
@@ -0,0 +1,255 @@
+bound value is static.
+
+The following type represents a 2D tensor, with the size of the 0th dimension
+being up to 3 and the size of the 1st dimension being exactly 5:
+
+```mlir
+tensor<?x5xf32, #stablehlo.type_extensions<bounds = [3, ?]>>
+```
+
+Type compatibility in StableHLO also checks for compatibility of the bounds.
+For example, the example type above is compatible with `tensor<2x5xf32>` but not
+with `tensor<7x5xf32>` as it doesn’t respect the bound `3` on the first
+dimension.
+
+```mlir
+func.func @bounds_compatibility(%arg0: tensor<?xf32, #stablehlo.type_extensions<bounds = [3]>>,
+                                %arg1: tensor<?xf32, #stablehlo.type_extensions<bounds = [2]>>,
+                                %arg2: tensor<?xf32, #stablehlo.type_extensions<bounds = [4]>>,
+                                %arg3: tensor<2xf32>,
+                                %arg4: tensor<4xf32>) {
+  // %arg0 is compatible with %arg1, %arg2 and %arg3 as bounded types could have
+  // tensor<2xf32> type during runtime.
+  %0 = "stablehlo.add"(%arg0, %arg1) : (tensor<?xf32, #stablehlo.type_extensions<bounds = [3]>>, tensor<?xf32, #stablehlo.type_extensions<bounds = [2]>>) -> tensor<?xf32, #stablehlo.type_extensions<bounds = [2]>>
+  %1 = "stablehlo.add"(%arg0, %arg2) : (tensor<?xf32, #stablehlo.type_extensions<bounds = [3]>>, tensor<?xf32, #stablehlo.type_extensions<bounds = [4]>>) -> tensor<?xf32, #stablehlo.type_extensions<bounds = [3]>>
+  %2 = "stablehlo.add"(%arg0, %arg3) : (tensor<?xf32, #stablehlo.type_extensions<bounds = [3]>>, tensor<2xf32>) -> tensor<2xf32>
+
+  // This is illegal as operands have incompatible types. %arg0 can either be
+  // tensor<0xf32>, tensor<1xf32>, tensor<2xf32> or tensor<3xf32> at runtime,
+  // none of these are compatible with tensor<4xf32>
+  %3 = "stablehlo.add"(%arg0, %arg4) : (tensor<?xf32, #stablehlo.type_extensions<bounds = [3]>>, tensor<4xf32>) -> tensor<*xf32>
+  func.return
+}
+```
+
+Currently, the StableHLO dialect uses the MLIR ranked tensor type to represent
+bounds. In the future we plan to introduce StableHLO type which supports bounds,
+along with a custom pretty printing format.  There will be a separate RFC on
+this. Also, the proposal will follow StableHLO backward compatibility policies
+so it is safe to use `TypeExtensionsAttr` currently.
+
+### (P2) StableHLO op semantics with bounded operands or results
+
+All ops that support dynamic operands or results can have bounds specified for
+them. However, the result types need to be compatible with the inferred result
+types. This allows result types to be more generic or specific as long as it is
+compatible with the inferred type.
+
+Separately, the StableHLO specification will be updated to cover bounded types
+for all the relevant ops.
+
+### (P3) get\_dimension\_size / set\_dimension\_size ops
+
+The `get_dimension_size` op takes a tensor and a dimension index and returns the
+runtime size as `tensor<i32>`.
```

Comment:
Any reason to not use `tensor<index>`?

Reply:
I was thinking of generalizing to i64 and index as well for consistency. But given there is no requirement now. Seemed better to follow XLA until some use-case comes up.